database.go

```diff
@@ -166,3 +164,71 @@ func (e ErrRequiredAttributeMissing) Error() string {
 		string(e),
 	)
 }
+
+type BinariesUsages map[int][]*BinaryReference
+
+func (db *Database) getBinaries() *Binaries {
+	if db.Header.IsKdbx4() {
+		return &db.Content.InnerHeader.Binaries
+	}
+
+	return &db.Content.Meta.Binaries
+}
+
+func (db *Database) cleanupBinaries() int {
+	usages := db.getBinariesUsages()
+	binaries := *db.getBinaries()
+	updated := Binaries{}
+	removed := 0
+	counter := 0
+
+	for _, binary := range binaries {
+		if _, ok := usages[binary.ID]; ok {
+			for _, ref := range usages[binary.ID] {
+				ref.Value.ID = counter
+			}
+			binary.ID = counter
+			updated = append(updated, binary)
+			counter++
+		} else {
+			removed++
+		}
+	}
+
+	*db.getBinaries() = append(Binaries{}, updated...)
```

Comment:
I would prefer to avoid this bytes extra copy and edit them in place, but not sure how to implement it in Go.

Reply:
Since binaries is not just a list of pointers, any editing of the list necessarily results in copying of data.

Even if you could directly work on the binaries slice, removing something from it would require copying.

Unless you have KDBX files with lots of attached files, I'd say: go for the dumb version.
Something that any average programmer without a background in e.g. systems programming (where such optimizations are relevant) would understand.

Most of the time, the simpler code is good enough and a lot easier to maintain.

The part that you could optimize though is not running the append here at all and instead assign `updated`